### PR TITLE
feat: Kubernetes Operator 

### LINKED ISSUE #138: Kubernetes Controller
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

We need an interaction layer with Kubernetes, this should likely be done as a Kubernetes controller so that we can interact with CRDs and follow the operator pattern.

### Describe the solution you'd like

We need to get the existing KubeOps branch merged into main, this branch contains some database CRDs and work.

### Additional context

On top of getting the branch mergable, we should add an Application CRD as well. This should contain the following fields:

- Application Name
- Application Description

### PR DESCRIPTION
**Pull request by babanfaraj**
# Add KubeOps 

- [X] You've read the [Contributor Guide](https://github.com/recro/medulla/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/recro/medulla/blob/main/CODE_OF_CONDUCT.md).
- [X] You've included unit or integration tests for your change, where applicable.
- [X] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Summary of the changes

This changes contains added source code for KubeOps.

## Description

The files added are regarding KubeOps which contains entities, CRDs and other C# files. Almost all the source code was written by @keithlogan94

Fixes #138

### CODE REVIEW

File: src/Documentation/Client/Pages/Documentation.razor [hunk 1]
```diff
@@ -5,6 +5,7 @@
 @page "/"
 @page "/{category}"
 @page "/{category}/{page}"
+@using System.IO
```
Comment (AndrewBabbitt97) [inline src/Documentation/Client/Pages/Documentation.razor:8]: Lets keep unrelated changes in separate PRs so they can be reverted if needed.

File: src/Documentation/Client/Pages/Search.razor [hunk 2]
```diff
@@ -3,6 +3,7 @@
 @* See the LICENSE file in the project root for more information. *@
 
 @page "/search"
+@using System.IO
```
Comment (AndrewBabbitt97) [inline src/Documentation/Client/Pages/Search.razor:6]: Lets keep unrelated changes in separate PRs so they can be reverted if needed.

File: src/Documentation/Client/Shared/MainLayout.razor [hunk 3]
```diff
@@ -3,6 +3,7 @@
 @* See the LICENSE file in the project root for more information. *@
 
 @using System.Web
+@using System.IO
```
Comment (AndrewBabbitt97) [inline src/Documentation/Client/Shared/MainLayout.razor:6]: Lets keep unrelated changes in separate PRs so they can be reverted if needed.

File: src/Medulla/Kubernetes/Operator/Operator.csproj.DotSettings [hunk 4]
```diff
@@ -0,0 +1,2 @@
+<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
+	<s:String x:Key="/Default/CodeInspection/CSharpLanguageProject/LanguageLevel/@EntryValue">CSharp100</s:String></wpf:ResourceDictionary>
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Operator.csproj.DotSettings:2]: Not sure why we have a WPF resource dictionary here, is this actually needed? If so why? If not please remove.

File: src/Medulla/Kubernetes/Operator/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddKubernetesOperator();
+    }
+
+    public void Configure(IApplicationBuilder app)
+    {
+        app.UseKubernetesOperator();
+    }
+}
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Startup.cs:20]: This is the old method of bootstrapping an ASP.NET app, lets move all this logic to [top level statements](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/top-level-statements) inside of Program.cs

You can see how this is done in other Program.cs files in the repo.

By doing this this entire file should end up deleted.

File: src/Medulla/Kubernetes/Operator/Program.cs [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    /// Main
+    /// </summary>
+    /// <param name="args"></param>
+    /// <returns></returns>
+    public static Task<int> Main(string[] args)
+    {
+        Console.WriteLine("Trying to get uri for database service");
+        Console.WriteLine("List Service Uris: ");
+        Console.WriteLine("Service Uri {0} is {1}", "database service", ServiceUris.GetDatabaseServiceUri());
+        Console.WriteLine("Service Uri {0} is {1}", "sync service", ServiceUris.GetDatabaseSyncUri());
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Program.cs:24]: This can fail depending on runtime mode, Console is not always available. Use the [.NET Logger](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging/?view=aspnetcore-6.0) if needed. Generally avoid logger output in Program.cs

File: src/Medulla/Kubernetes/Operator/Operator.csproj [hunk 7]
```diff
@@ -0,0 +1,24 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <UserSecretsId>ef6be8d2-d4c9-41af-92be-9e15a3a962ea</UserSecretsId>
+    <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+    <LangVersion>preview</LangVersion>
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Operator.csproj:9]: Remove these lines, they are either inherited from build properties or just things we don't want left in.

File: src/Medulla/Kubernetes/Operator/Program.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+        Console.WriteLine("GetEnvironmentVariables: ");
+        foreach (DictionaryEntry de in Environment.GetEnvironmentVariables())
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Program.cs:27]: Don't read in stuff from Environment, use [.NET Configuration](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-6.0) instead.

File: src/Medulla/Kubernetes/Operator/Program.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    private static IHostBuilder CreateHostBuilder(string[] args)
+    {
+        return Host.CreateDefaultBuilder(args)
+            .ConfigureWebHostDefaults(webBuilder =>
+            {
+                webBuilder.UseStartup<Startup>();
+            });
+    }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Program.cs:44]: This is the old method of bootstrapping an ASP.NET app, lets move all this logic to [top level statements](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/top-level-statements) inside of Program.cs

You can see how this is done in other Program.cs files in the repo.

File: src/Medulla/Kubernetes/Operator/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+RUN addgroup k8s-operator && useradd -G k8s-operator operator-user
+
+WORKDIR /operator
+COPY --from=build /operator/out/ ./
+RUN chown operator-user:k8s-operator -R .
+
+USER operator-user
+
+ENTRYPOINT [ "dotnet", "Operator.dll" ]
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Dockerfile:19]: Entire file should be able to be deleted, Tye handles containerization of the application.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    public bool AllowNull { get; set; }
+    public string? DefaultValue { get; set; }
+    public bool AutoIncrement { get; set; }
+    public bool PrimaryKey { get; set; }
+    public string? Field { get; set; }
+    public string? Unique { get; set; }
+    public string? Comment { get; set; }
+    public ValidateSpec? Validate { get; set; }
+
+}
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs:85]: Avoid sticking multiple structs / classes in a single file.

If you have a struct named `ColumnSpec` it should be in a file named `ColumnSpec.cs`. Its very common to reuse structs across multiple CRDs so we shouldn't group them under a single file.

We are also missing XML comments for all the public members here.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+/// <summary>
+/// V1Alpha1DataEntity is CRD for data
+/// </summary>
+[Description("A CustomResourceDefinition which allows building with data in Medulla.")]
+[KubernetesEntity(
+    ApiVersion = "v1alpha1",
+    Kind = "Data",
+    Group = "medulla.recro.com",
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs:94]: Group should be `medulla.io`

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 13]
```diff
@@ -0,0 +1,99 @@
+﻿// Licensed to the Medulla Contributors under one or more agreements.
+// The Medulla Contributors licenses this file to you under the Apache 2.0 license.
+// See the LICENSE file in the project root for more information.
+
+using k8s.Models;
+using KubeOps.Operator.Entities;
+using KubeOps.Operator.Entities.Annotations;
+
+namespace DatabaseControllerKubeOps.Controller.Entities;
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs:9]: Namespace should be `Medulla.Kubernetes.Operator.Entities`

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+using KubeOps.Operator.Entities;
+using KubeOps.Operator.Entities.Annotations;
+
+namespace ApplicationControllerKubeOps.Controller.Entities;
+
+
+/// <summary>
+/// ApplicationSpec is spec of CRD for application
+/// </summary>
+public struct ApplicationSpec
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs:15]: Missing XML comments

**Review by AndrewBabbitt97 — CHANGES_REQUESTED** [review]
Lets start with these changes, it should be fairly close if not good after.

File: src/Medulla/Kubernetes/Operator/Operator.csproj.DotSettings (hunk 4 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Operator.csproj.DotSettings:2]: @keithlogan94  was this added for the database service? I can remove it from this PR.

File: src/Medulla/Kubernetes/Operator/Startup.cs (hunk 5 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Startup.cs]: @AndrewBabbitt97  Just pushed up the latest changes regarding this by following this documentation here: https://docs.microsoft.com/en-us/aspnet/core/migration/50-to-60?view=aspnetcore-6.0&tabs=visual-studio. Let me know if I missed something.

Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Startup.cs]: Marking this as resolved unless further updates are needed.

File: src/Medulla/Kubernetes/Operator/Program.cs (hunk 6 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Program.cs]: Removed

File: src/Medulla/Kubernetes/Operator/Operator.csproj (hunk 7 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Operator.csproj]: Removed

File: src/Medulla/Kubernetes/Operator/Program.cs (hunk 8 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Program.cs]: Decided to remove this anyways unless you think we need it @keithlogan94 ?

File: src/Medulla/Kubernetes/Operator/Program.cs (hunk 9 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Program.cs]: Resolved. Please see the other thread we had regarding this above and we can track further requests there too.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs (hunk 14 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs:15]: @AndrewBabbitt97  @keithlogan94  how is this missing XML comments? Is it not detailed enough or this file does not contain the desired XML comment? I used V1Alpha1DataEntity.cs as an example.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs (hunk 11 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs]: Moved struct usage into a new model/database directory:  https://github.com/recro/medulla/pull/153/commits/a402498366570dc2946880699739e0fabef958b5

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs (hunk 13 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs]: Resolved.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs (hunk 14 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs:15]: Added XML comments.

File: src/Medulla/Kubernetes/Operator/model/database/ColumnSpec.cs [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    /// </summary>
+    public struct ColumnSpec
+    {
+        public string? ColumnName { get; set; }
+        public string? Type { get; set; }
+        public bool AllowNull { get; set; }
+        public string? DefaultValue { get; set; }
+        public bool AutoIncrement { get; set; }
+        public bool PrimaryKey { get; set; }
+        public string? Field { get; set; }
```
Comment (babanfaraj) [inline src/Medulla/Kubernetes/Operator/model/database/ColumnSpec.cs:18]: @keithlogan94  Why do we have a string field in here? Is this a string that contains all fields of the corresponding column? https://docs.microsoft.com/en-us/office/vba/outlook/concepts/forms/difference-between-a-field-and-a-column

Comment (keithlogan94) [reply src/Medulla/Kubernetes/Operator/model/database/ColumnSpec.cs:18]: @babanfaraj ansered in google chat

File: src/Medulla/Kubernetes/Operator/ServiceUris.cs [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    public static string GetDatabaseSyncUri()
+    {
+        return $"http://{GetEnvNotEmpty("SERVICE__DATABASE-SYNC__HTTP__HOST")}:{GetEnvNotEmpty("SERVICE__DATABASE-SYNC__HTTP__PORT")}";
+    }
+
+
+
+
+
+}
```
[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/ServiceUris.cs:42]: @AndrewBabbitt97  may remove this from this PR as these methods are not being used in the source code in this PR. I believe these are used with the database service and that is where it should be updated. Awaiting on @keithlogan94 response to see what to do.

File: src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs [hunk 17]
```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the Medulla Contributors under one or more agreements.
+// The Medulla Contributors licenses this file to you under the Apache 2.0 license.
+// See the LICENSE file in the project root for more information.
+
+namespace Operator.Model
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:5]: Namespace should be Medulla.Kubernetes.Operator.Model.Database

Use [file scoped namespaces](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces) instead of curly brace oriented ones

File: src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs [hunk 18]
```diff
@@ -0,0 +1,26 @@
+﻿// Licensed to the Medulla Contributors under one or more agreements.
+// The Medulla Contributors licenses this file to you under the Apache 2.0 license.
+// See the LICENSE file in the project root for more information.
+
+using KubeOps.Operator.Entities.Annotations;
+
+namespace Operator.Model
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs:7]: Namespace should be Medulla.Kubernetes.Operator.Model.Database

Use [file scoped namespaces](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces) instead of curly brace oriented ones

File: src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+    /// <summary>
+    /// Spec of CRD for model/table within Database
+    /// </summary>
+    public struct ModelSpec
+    {
+        /// <summary>
+        /// Name is an element of ModelSpec which is the name of the model/table
+        /// </summary>
+        [Required]
+        public string? Name { get; set; }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs:18]: Name should not be nullable, as its a required field.

File: src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+        /// </summary>
+        public string? Is { get; set; }
+        /// <summary>
+        /// matches this RegExp /^[a-z]+$/i
+        /// </summary>
+        public string? Not { get; set; }
+        /// <summary>
+        /// checks for email format (foo@bar.com)
+        /// </summary>
+        public bool? IsEmail { get; set; }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:24]: Bools have a default value of false, I think this is appropriate for this and the following values? @keithlogan94 if you want to chime in

File: src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+        /// <summary>
+        /// Columns is an element of ModelSpec which contains a list of ColumnSpec of the model/table
+        /// </summary>
+        [Required]
+        public List<ColumnSpec>? Columns { get; set; }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs:23]: List should be not nullable, meaning it can be at least an empty list

File: src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        public bool AllowNull { get; set; }
+        /// <summary>
+        /// DefaultValue is an element of ColumnSpec which should be set to what the default value of the column should
+        /// be if it exists
+        /// </summary>
+        public string? DefaultValue { get; set; }
+        /// <summary>
+        /// AutoIncrement is an element of ColumnSpec which indicates whether auto increment is allowed or not
+        /// </summary>
+        public bool AutoIncrement { get; set; }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs:32]: All these bools should be set to false by default, no need for nullability. Double check this @keithlogan94

File: src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+{
+    /// <summary>
+    /// Spec of Column within Model as array
+    /// </summary>
+    public struct ColumnSpec
+    {
+        /// <summary>
+        /// ColumnName is an element of ColumnSpec which is the column name
+        /// </summary>
+        public string? ColumnName { get; set; }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs:15]: Columns must have a name; hence they cant be null

File: src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+        /// <summary>
+        /// Type is an element of ColumnSpec which is the column type
+        /// </summary>
+        public string? Type { get; set; }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs:19]: Columns must have a type, hence it cant be null

File: src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs [hunk 25]
```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the Medulla Contributors under one or more agreements.
+// The Medulla Contributors licenses this file to you under the Apache 2.0 license.
+// See the LICENSE file in the project root for more information.
+
+namespace Operator.Model
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ColumnSpec.cs:5]: Namespace should be Medulla.Kubernetes.Operator.Model.Database

Use [file scoped namespaces](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/file-scoped-namespaces) instead of curly brace oriented ones

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+using KubeOps.Operator.Entities.Annotations;
+using Operator.Model;
+
+namespace Medulla.Kubernetes.Operator.Entities;
+
+
+/// <summary>
+/// DatabaseSpec is spec of CRD for database
+/// </summary>
+public struct DatabaseSpec
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs:16]: Struct should be in its own file called DatabaseSpec.cs

**Review by AndrewBabbitt97 — CHANGES_REQUESTED** [review]
Make sure all fields have appropriate nullability and default values correct if not null, this is just about there. Mostly simple fixes.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+{
+    /// <summary>
+    /// Name is an element of DatabaseSpec which is the name of the database
+    /// </summary>
+    [Required]
+    public string? Name { get; set; }
```
[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs]: Agreed. Removing allowing null along with the rest of the attributes. I do not think any should be allowed to be null.

File: src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs (hunk 20 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:24]: So no need to change it then correct?

**Comment by babanfaraj** [comment]
@AndrewBabbitt97  ready for review again.

File: src/Medulla/Kubernetes/Operator/Model/Database/Database.cs [hunk 28]
```diff
@@ -25,7 +25,7 @@ public struct DatabaseSpec
     /// Dialect is an element of DatabaseSpec which is the the dialect of the database
     /// </summary>
     [Required]
-    public string? Dialect { get; set; }
+    public string Dialect { get; set; }
```
[the remark this answers is not in the document]
Comment (keithlogan94) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs:28]: @babanfaraj

File: src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs [hunk 29]
```diff
@@ -16,10 +16,10 @@ public struct ApplicationSpec
...
     /// </summary>
     [Required]
-    public string? Name { get; set; }
+    public string Name { get; set; }
     /// <summary>
     /// Description is is an element of the ApplicationSpec which is the description of the Application created by Medulla
     /// </summary>
     [Required]
-    public string? Description { get; set; }
+    public string Description { get; set; }
```
Comment (keithlogan94) [inline src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs:24]: Needs to be set immediately too

File: src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs [hunk 30]
```diff
@@ -16,10 +16,10 @@ public struct ApplicationSpec
     /// Name is is an element of the ApplicationSpec which is the name of the Application created by Medulla
     /// </summary>
     [Required]
-    public string? Name { get; set; }
+    public string Name { get; set; }
```
[the remark this answers is not in the document]
Comment (keithlogan94) [reply src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs:19]: or = "";

File: src/Medulla/Kubernetes/Operator/Model/Database/Database.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs:28]: Why are we requiring default values on the variable Dialect and not the other variables in Database.cs?

[the remark this answers is not in the document]
Comment (AndrewBabbitt97) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs:28]: Tbh this is more of an internal detail, and the reality is users shouldnt need to worry about this at all for the time being

File: src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs (hunk 30 above)
[the remark this answers is not in the document]
Comment (AndrewBabbitt97) [reply src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs:19]: Should be string.Empty

File: src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs (hunk 29 above)
Comment (AndrewBabbitt97) [reply src/Medulla/Kubernetes/Operator/Model/Application/ApplicationSpec.cs:24]: string.Empty

File: src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs (hunk 21 above)
Comment (AndrewBabbitt97) [reply src/Medulla/Kubernetes/Operator/Model/Database/ModelSpec.cs]: Still not resolved, the list should not be null

File: src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs [hunk 31]
```diff
@@ -0,0 +1,95 @@
...
+/// <summary>
+/// Spec in Column for validation
+/// </summary>
+public struct ValidateSpec
+{
+
+    /// <summary>
+    /// ValidateSpec Constructor
+    /// </summary>
+    public ValidateSpec()
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:16]: Can be deleted, no need for an empty constructor

File: src/Medulla/Kubernetes/Operator/Model/Database/Database.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs:28]: Alright, it will not be required and setting default to mysql.

**Review by AndrewBabbitt97 — CHANGES_REQUESTED** [review]

File: src/Medulla/Kubernetes/Operator/Model/Database/Database.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (keithlogan94) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs:28]: @babanfaraj The reason for requiring default values, is basically enforcing something is not null.
If you set a variable to a value then it's not null, and can't produce a null exception.

[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs:28]: @keithlogan94  I was not asking the reason for requiring default values. I was asking why we should not have default values on the other elements in Database.cs besides the Dialect element.

File: src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs (hunk 31 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:16]: I believe @keithlogan94  added that there because we do need a constructor when we add default values. Please see the following error once I remove the empty constructor:

`
PS C:\Users\Baban\medulla\medulla\src\Medulla\Kubernetes\Operator> dotnet run
C:\Users\Baban\medulla\medulla\src\Medulla\Kubernetes\Operator\Model\Database\ValidateSpec.cs(10,15): error CS8983: A 'struct' with field initializers must include an explicitly declared constructor. [C:\Users\Baban\medulla\medulla\src\Medulla\Kubernetes\Operator\Operator.csproj]

The build failed. Fix the build errors and run again.
`

Comment (keithlogan94) [reply src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:16]: ValidateSpec.cs(10,15): error CS8983: A 'struct' with field initializers must include an explicitly declared constructor.

Comment (keithlogan94) [reply src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:16]: @AndrewBabbitt97 I added it because of this error

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+    ApiVersion = "v1alpha1",
+    Kind = "Data",
+    Group = "medulla.io",
+    PluralName = "data")]
+public class V1Alpha1DataEntity : CustomKubernetesEntity
+{
+    /// <summary>
+    /// Spec is a list elements of type DatabaseSpec
+    /// </summary>
+    public List<DatabaseSpec>? Spec { get; set; }
```
[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs]: Resolved. Created a list of Database Spec.

File: src/Medulla/Kubernetes/Operator/Model/Database/Database.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/Database.cs]: I added default values for the rest of the elements.

File: src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs (hunk 31 above)
Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:16]: I also followed the same pattern in the other files in this PR.

Comment (babanfaraj) [reply src/Medulla/Kubernetes/Operator/Model/Database/ValidateSpec.cs:16]: I assume we can keep this right @AndrewBabbitt97  ?  If so, all other comments have been addressed too.

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs [hunk 33]
```diff
@@ -0,0 +1,30 @@
...
+/// </summary>
+[Description("A CustomResourceDefinition which allows building with data in Medulla.")]
+[KubernetesEntity(
+    ApiVersion = "v1alpha1",
+    Kind = "Application",
+    Group = "medulla.io",
+    PluralName = "application")]
+public class V1Alpha1ApplicationEntity : CustomKubernetesEntity
+{
+    public V1Alpha1ApplicationEntity() { }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1ApplicationEntity.cs:24]: Remove empty class constructor or document it

File: src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs [hunk 34]
```diff
@@ -0,0 +1,30 @@
...
+/// </summary>
+[Description("A CustomResourceDefinition which allows building with data in Medulla.")]
+[KubernetesEntity(
+    ApiVersion = "v1alpha1",
+    Kind = "Data",
+    Group = "medulla.io",
+    PluralName = "data")]
+public class V1Alpha1DataEntity : CustomKubernetesEntity
+{
+    public V1Alpha1DataEntity() { }
```
Comment (AndrewBabbitt97) [inline src/Medulla/Kubernetes/Operator/Entities/V1Alpha1DataEntity.cs:24]: Remove empty class constructor or document it

**Review by AndrewBabbitt97 — CHANGES_REQUESTED** [review]

**Comment by AndrewBabbitt97** [comment]
Simple two changes, not sure why these were added after the previous review as they were never requested to be added

**Comment by AndrewBabbitt97** [comment]
Otherwise LGTM

**Comment by babanfaraj** [comment]
> Simple two changes, not sure why these were added after the previous review as they were never requested to be added

I believe I added it at the time to allow default values element in the associated class.

**Review by AndrewBabbitt97 — APPROVED** [review]
